File: Ahorn/entities/TitanThrone.jl
module TitanThrone

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/TitanThrone" TitanThrone(x::Integer, y::Integer, activationRadius::Integer=100, autoActivate::Bool=false, bossEntity::String="MaggyHelper/KingTitanBoss", cutsceneId::String="CH15_ROARING_TITAN_KING_BATTLE")

const placements = Ahorn.PlacementDict(
    "manual" => Ahorn.EntityPlacement(TitanThrone),
    "auto" => Ahorn.EntityPlacement(TitanThrone)
)

function Ahorn.selection(entity::TitanThrone)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::TitanThrone, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: Loenn/entities/titanThrone.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local titanThrone = {}

titanThrone.name = "MaggyHelper/TitanThrone"
titanThrone.depth = -100
titanThrone.fieldInformation = {
    activationRadius = {
        minimumValue = 0
    }
}
titanThrone.placements = {
    {
        name = "manual",
        data = {
            bossEntity = "MaggyHelper/KingTitanBoss",
            activationRadius = 100,
            autoActivate = false,
            cutsceneId = "CH15_ROARING_TITAN_KING_BATTLE"
        }
    },
    {
        name = "auto",
        data = {
            bossEntity = "MaggyHelper/KingTitanBoss",
            activationRadius = 100,
            autoActivate = true,
            cutsceneId = "CH15_ROARING_TITAN_KING_BATTLE"
        }
    }
}

function titanThrone.sprite(room, entity)
    local sprite = drawableSprite.fromTexture("objects/titan_throne/inactive", entity)
    sprite:setPosition(entity.x, entity.y)
    sprite:setJustification(0.5, 1.0)
    sprite.color = {0.6, 0.5, 0.4, 1.0}
    return sprite
end

function titanThrone.selection(room, entity)
    return utils.rectangle(entity.x - 40, entity.y - 120, 80, 120)
end

return titanThrone
